### PR TITLE
[java] Support Java 19

### LINKED ISSUE #4015: [java] Support JDK 19
[OpenJDK 19](https://openjdk.java.net/projects/jdk/19/)  GA is scheduled for 2022/09/20.
So we should add support for that with PMD version 6.49.0 or earlier.

The following JEPs are important for us:
*   [JEP 405: Record Patterns (Preview)](https://openjdk.java.net/jeps/405)
    * destructuring of record components with instanceof
*   [JEP 427: Pattern Matching for switch (Third Preview)](https://openjdk.java.net/jeps/427)
    * new keyword `when`...

In summary: https://javaalmanac.io/jdk/19/

This issue is similar to the previous: #3809 

EA Release of JDK19 is available here: http://jdk.java.net/19/

**Tasks:**

- [ ] Add new java language versions 19 and 19-preview.
    - 19 is the new default
    - update also java.md and ant.md
    - and add in KotlinTestingDsl.kt.
- [ ] Update asm dependency:
    - https://search.maven.org/artifact/org.ow2.asm/asm
    - https://asm.ow2.io/versions.html
    - Update net.sourceforge.pmd.lang.java.typeresolution.visitors.PMDASMVisitor.ASM_API if necessary
- [ ] Update implementation for "Pattern Matching for switch" (406, 420, 427) as a preview feature, mark AST classes as experimental, if needed.
    - https://openjdk.java.net/jeps/406 (preview 1)
    - https://openjdk.java.net/jeps/420 (preview 2)
    - https://openjdk.java.net/jeps/427 (preview 3)
    - http://cr.openjdk.java.net/~gbierman/jep427%2b405/jep427+405-20220517/specs/patterns-switch-record-patterns-jls.html
- [ ] Implement "Record Patterns" (405) as a preview feature, mark AST classes as experimental, if needed.
    - https://openjdk.java.net/jeps/405
    - http://cr.openjdk.java.net/~gbierman/jep427%2b405/jep427+405-20220517/specs/patterns-switch-record-patterns-jls.html
- [ ] Remove java language version 17-preview
    - update also java.md and ant.md
    - update KotlinTestingDsl.kt
- [ ] Write release notes
- [ ] Prepare merge for PMD 7 as a new PR

### PR DESCRIPTION
**Pull request by adangel**
## Describe the PR

~Note: I didn't prepare yet a merge to pmd7, that's still on the TODO list...~

For pmd7, see #4039 

## Related issues

- Fixes #4015 

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Review by oowekyala — APPROVED** [review]
Thanks for doing this, looks good! I can take on the merge to PMD 7 once this is merged if you like

File: pmd-java/etc/grammar/Java.jjt [hunk 1]
```diff
@@ -1785,15 +1811,35 @@ void TypePattern():
   VariableDeclaratorId()
 }
 
+void RecordPattern():
+{ checkForRecordPatterns(); }
+{
+  ReferenceType() RecordStructurePattern() [ VariableDeclaratorId() ]
+}
+
+void RecordStructurePattern():
```
[the remark this answers is not in the document]
Comment (adangel) [reply pmd-java/etc/grammar/Java.jjt:1820]: :+1:

File: pmd-java/etc/grammar/Java.jjt [hunk 2]
```diff
@@ -2285,6 +2331,21 @@ void CaseLabelElement(ASTSwitchLabel label) #void:
         }
 }
 
+void Guard() :
```
[the remark this answers is not in the document]
Comment (adangel) [reply pmd-java/etc/grammar/Java.jjt:2334]: Let's go for ASTSwitchGuard for now

**Comment by adangel** [comment]
Thanks for the review!
I'll try to the merge to PMD7 myself first :)

**Comment by oowekyala** [comment]
Merged, special thanks for the nice documentation!

**Pull request merged**